server/storage/couchdb.go

```diff
@@ -0,0 +1,591 @@
+		Version:        a.Version,
+		SHA256:         a.SHA256,
+		Data:           a.Data,
+		TSchecksum:     a.TSchecksum,
+	}, nil
+}
+
+func cdbChangeFromJSON(data []byte) (interface{}, error) {
+	res := CChange{}
+	if err := json.Unmarshal(data, &res); err != nil {
+		return CChange{}, err
+	}
+	return res, nil
+}
+
+// CouchDB implements a MetaStore against the Couchdb Database
+type CouchDB struct {
+	dbName   string
+	client   *kivik.Client
+	user     string
+	password string
+}
+
+// NewCouchDBStorage initializes a CouchDB object
+func NewCouchDBStorage(dbName, user, password string, client *kivik.Client) CouchDB {
+	return CouchDB{
+		dbName:   dbName,
+		client:   client,
+		user:     user,
+		password: password,
+	}
+}
+
+func genID(gun, role string, version int) string {
+	return fmt.Sprintf("%s.%s.%d", gun, role, version)
+}
+
+// UpdateCurrent adds new metadata version for the given GUN if and only
+// if it's a new role, or the version is greater than the current version
+// for the role. Otherwise an error is returned.
+func (cdb CouchDB) UpdateCurrent(gun data.GUN, update MetaUpdate) error {
+
+	file := CDBTUFFile{}
+
+	db, err := couchdb.DB(cdb.client, cdb.dbName, file.TableName())
+	if err != nil {
+		return err
+	}
+
+	id := genID(gun.String(), update.Role.String(), update.Version)
+
+	row, _ := db.Get(context.TODO(), id)
+	if row != nil {
+		return ErrOldVersion{}
+	}
+	// empty string is the zero value for tsChecksum in the CDBTUFFile struct.
+	// Therefore we can just call through to updateCurrentWithTSChecksum passing
+	// "" for the tsChecksum value.
+	if err := cdb.updateCurrentWithTSChecksum(gun.String(), "", update); err != nil {
+		return err
+	}
+	if update.Role == data.CanonicalTimestampRole {
+		tsChecksumBytes := sha256.Sum256(update.Data)
+		return cdb.writeChange(
+			gun.String(),
+			update.Version,
+			hex.EncodeToString(tsChecksumBytes[:]),
+			changeCategoryUpdate,
+		)
+	}
+	return nil
+}
+
+// updateCurrentWithTSChecksum adds new metadata version for the given GUN with an associated
+// checksum for the timestamp it belongs to, to afford us transaction-like functionality
+func (cdb CouchDB) updateCurrentWithTSChecksum(gun, tsChecksum string, update MetaUpdate) error {
+	now := time.Now()
+	checksum := sha256.Sum256(update.Data)
+
+	id := genID(gun, update.Role.String(), update.Version)
+
+	file := CDBTUFFile{
+		Timing: couchdb.Timing{
+			CreatedAt: now,
+			UpdatedAt: now,
+		},
+		ID: couchdb.ID{
+			ID: id,
+		},
+		GunRoleVersion: []interface{}{gun, update.Role, update.Version},
+		Gun:            gun,
+		Role:           update.Role.String(),
+		Version:        update.Version,
+		SHA256:         hex.EncodeToString(checksum[:]),
+		TSchecksum:     tsChecksum,
+		Data:           update.Data,
+	}
+
+	db, err := couchdb.DB(cdb.client, cdb.dbName, file.TableName())
+	if err != nil {
+		return err
+	}
+
+	_, err = db.Put(context.TODO(), id, file)
+	if err != nil {
+		return ErrOldVersion{}
```

Comment:
Hmm... more error parsing may be needed.  It seems like there may be other reasons a `db.Put` would error other than a conflict, which is what would signify an old version.

Reply:
I am parsing the returned error string now and return err if it doesn't match, ErrOldVersion() otherwise.